File: app/src/main/java/widget/CounterListAdapter.java
package widget;

import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ArrayAdapter;
import android.widget.TextView;

import java.util.List;

import asvid.beercounter.R;
import asvid.beercounter.data.CounterItem;

/**
 * Created by adam on 15.01.17.
 */
public class CounterListAdapter extends ArrayAdapter<CounterItem> {

    private final int viewResourceId;
    List<CounterItem> items;

    @Override
    public int getCount() {
        return items.size();
    }

    @Override
    public long getItemId(int i) {
        return items.get(i).getId();
    }

    public CounterListAdapter(Context context, int viewResourceId,
                              List<CounterItem> items) {
        super(context, viewResourceId, items);
        this.items = items;
        this.viewResourceId = viewResourceId;
    }

    public View getView(int position, View convertView, ViewGroup parent) {
        View v = convertView;
        if (v == null) {
            LayoutInflater vi = (LayoutInflater) getContext()
                    .getSystemService(Context.LAYOUT_INFLATER_SERVICE);
            v = vi.inflate(viewResourceId, null);
        }
        CounterItem item = items.get(position);
        TextView name = (TextView) v.findViewById(R.id.name);
        TextView value = (TextView) v.findViewById(R.id.value);
        name.setText(item.getName());
        value.setText(item.getValue().toString());
        return v;
    }
}

File: app/src/main/java/asvid/beercounter/CounterListAdapter.kt
package asvid.beercounter

import android.support.v7.widget.RecyclerView
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import asvid.beercounter.data.CounterItem
import com.mikepenz.iconics.view.IconicsButton

/**
 * Created by adam on 15.01.17.
 */
class CounterListAdapter(private val items: MutableList<CounterItem>,
    private val listener: CounterListListener) : RecyclerView.Adapter<CounterListAdapter.CounterItemViewHolder>() {

    init {
        Log.d("CounterListAdapter", "list: " + items)
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): CounterItemViewHolder {
        val view = LayoutInflater.from(parent.context)
            .inflate(R.layout.counter_item_card, parent, false)
        return CounterItemViewHolder(view)
    }

    override fun onBindViewHolder(holder: CounterItemViewHolder, position: Int) {
        val item = items[position]
        Log.d("CounterListAdapter", "item: " + item)
        if (item.isValid) {
            holder.item = item
            holder.name.text = item.name
            holder.value.text = item.value.toString()

            holder.deleteButton.setOnClickListener { listener.onItemDelete(item) }
            holder.decrementButton.setOnClickListener { listener.onItemDelete(item) }
            holder.incrementButton.setOnClickListener { listener.onItemIncrement(item) }
        }
    }

    override fun getItemId(i: Int): Long {
        return items[i].id!!
    }

    override fun getItemCount(): Int {
        return items.size
    }

    fun addItem(counterItem: CounterItem) {
        items.add(counterItem)
        notifyItemInserted(itemCount - 1)
    }

    inner class CounterItemViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {

        var item: CounterItem? = null
        var name = itemView.findViewById(R.id.name) as TextView
        var value = itemView.findViewById(R.id.value) as TextView
        var deleteButton = itemView.findViewById(R.id.deleteButton) as IconicsButton
        var incrementButton = itemView.findViewById(R.id.incrementButton) as IconicsButton
        var decrementButton = itemView.findViewById(R.id.decrementButton) as IconicsButton
    }
}
